tests_e2e/tests/lib/add_network_security_group.py

```diff
@@ -0,0 +1,140 @@
+        #         "apiVersion": "2020-05-01",
+        #         "type": "Microsoft.Network/virtualNetworks",
+        #         "name": "[parameters('virtual_network_name')]",
+        #         "location": "[parameters('location')]",
+        #         "properties": {
+        #             "addressSpace": {
+        #                 "addressPrefixes": [
+        #                     "10.0.0.0/16"
+        #                 ]
+        #             },
+        #             "copy": [
+        #                 {
+        #                     "name": "subnets",
+        #                     "count": "[parameters('subnet_count')]",
+        #                     "input": {
+        #                         "name": "[concat(parameters('subnet_prefix'), copyIndex('subnets'))]",
+        #                         "properties": {
+        #                             "addressPrefix": "[concat('10.0.', copyIndex('subnets'), '.0/24')]"
+        #                         }
+        #                     }
+        #                 }
+        #             ]
+        #         }
+        #     }
+        #
+        network_resource = self._get_resource(resources, "Microsoft.Network/virtualNetworks")
+
+        # Add a dependency on the NSG
+        nsg_reference = f"[resourceId('Microsoft.Network/networkSecurityGroups', '{NETWORK_SECURITY_GROUP}')]"
+        network_resource_dependencies = network_resource.get("dependsOn")
+        if network_resource_dependencies is None:
+            network_resource["dependsOn"] = [nsg_reference]
+        else:
+            network_resource_dependencies.append(nsg_reference)
+
+        subnets_copy = network_resource["properties"].get("copy") if network_resource.get("properties") is not None else None
+        if subnets_copy is None:
+            raise Exception("Cannot find the copy property of the virtual network in the ARM template")
+
+        subnets = [i for i in subnets_copy if i["name"] == 'subnets']
+        if len(subnets) == 0:
+            raise Exception("Cannot find the subnets of the virtual network in the ARM template")
+
+        subnets_input = subnets[0].get("input")
+        if subnets_input is None:
+            raise Exception("Cannot find the input property of the subnets in the ARM template")
+
+        nsg_reference = json.loads(f"""{{
+          "networkSecurityGroup": {{
+            "id": "[resourceId('Microsoft.Network/networkSecurityGroups', '{NETWORK_SECURITY_GROUP}')]"
+          }}
+        }}""")
+
+        subnets_properties = subnets_input.get("properties")
+        if subnets_properties is None:
+            subnets_input["properties"] = nsg_reference
+        else:
+            subnets_properties.update(nsg_reference)
+
+    @property
+    def _my_ip_address(self) -> str:
+        """
+        Gets the IP address of the current machine.
+        """
+        if self.__my_ip_address is None:
+            connection = http.client.HTTPSConnection("ifconfig.io")
```

Comment:
This could fail with connection issue, should we handle it? so that we give a chance to run the pipeline without rule?

Reply:
yes, i should retry. 

skipping the rule on failure is a good suggestion. i'll do that as well